numba/_dispatcher.cpp

```diff
@@ -746,6 +759,290 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
     }
 }
 
+#elif (PY_MAJOR_VERSION >= 3) && (PY_MINOR_VERSION == 12)
+
+// Python 3.12 has a completely new approach to tracing and profiling due to
+// the new `sys.monitoring` system.
+
+// From: https://github.com/python/cpython/blob/0ab2384c5f56625e99bb35417cadddfe24d347e1/Python/instrumentation.c#L863-L868
+
+static const int8_t MOST_SIG_BIT[16] = {-1, 0, 1, 1,
+                                         2, 2, 2, 2,
+                                         3, 3, 3, 3,
+                                         3, 3, 3, 3};
+
+// From: https://github.com/python/cpython/blob/0ab2384c5f56625e99bb35417cadddfe24d347e1/Python/instrumentation.c#L873-L879
+
+static inline int msb(uint8_t bits) {
+    if (bits > 15) {
+        return MOST_SIG_BIT[bits>>4]+4;
+    }
+    return MOST_SIG_BIT[bits];
+}
+
+
+static int invoke_monitoring(PyThreadState * tstate, int event, Dispatcher *self, PyObject* retval)
```

Comment:
I've read this line by line and pondered the refcount. Looks ok.

Reply:
Thanks for checking.